1-Draft/RFCNNNN-Concurrent-Collections.md

```diff
@@ -0,0 +1,113 @@
+---
+RFC: RFCnnnn
+Author: Kirk Munro
+Status: Draft
+SupercededBy: 
+Version: 0.1
+Area: Parser
+Comments Due: July 25, 2019
+Plan to implement: Yes
+---
+
+# Closures for concurrent collections
+
+PowerShell is getting more and more support for multi-threaded use, such as:
+
+* the PSThreadJob module
+* `S.M.A.PowerShell.InvokeAsync` and `S.M.A.PowerShell.StopAsync`
+* parallelization in `ForEach-Object` ([see related RFC 194](https://github.com/PowerShell/PowerShell-RFC/pull/194))
+
+With these either available now or being made available soon, it would be
+helpful if scripters could easily create concurrent alternatives to arrays
+(`@()`) and hashtables (`@{}`).
+
+## Motivation
+
+    As a scripter,
+    I can easily define and work with a ConcurrentBag<PSObject>,
+    so that I can collect unordered objects from multiple threads in thread-safe manner.
+
+    As a scripter,
+    I can easily define and work with a ConcurrentDictionary<PSObject, PSObject>,
+    so that I can store objects relative to keys in a dictionary in thread-safe manner.
+
+## User Experience
+
+```powershell
+# Define a ConcurrentBag<PSObject>
+$collection = ~@()
+
+# Define a ConcurrentDictionary<PSObject,PSObject>
+$dictionary = ~@{}
+
+# Retrieve a bunch of logs from remote computers
+$computers = 's1','s2','s3'
+$computers | Start-ThreadJob {
+    # Retrieve the data and store it in a thread-safe collection
+    $collection += Get-LogData -ComputerName $_
+    # Retrieve the data and store it in a thread-safe dictionary
+    $dictionary[$_] = Get-LogData
+}
+
+# Process the collection as an array
+$collection.ToArray()
+
+# Process items in the dictionary
+foreach ($key in $dictionary.Keys) {
+    # Do something with each value
+    $dictionary[$key]
+}
+```
+
+```output
+The data gathered in the $collection collection.
+
+The data gathered in each key in the $dictionary collection.
+```
+
+## Specification
+
+* define `~@()` enclosures as `System.Collections.Concurrent.ConcurrentBag<PSObject>`
```

Comment:
There is a [related discussion](https://github.com/PowerShell/PowerShell/issues/5643) about supporting `System.Collections.Generic.List`
I think we need a unified approach otherwise we will get the Forth :-)

Reply:
Thank you for drawing my attention to that issue. I added comments over there and mentioned this RFC (you already made the link by referencing it, but I wanted to briefly explain the intent) so that the people thinking about this could consider the bigger picture.

Reply:
I also updated the RFC to specifically mention this discussion in the considerations section.